### ISSUE TITLE
SIGSEGV (code 134) during "Finalizing database" step

### ISSUE THREAD

**Issue by viktornordling** [github/codeql#14138]
We're analyzing a fairly large code base in a CircleCI Docker job. We've increased the resource class to the biggest one available (2xlarge+), but we're still seeing SIGSEGV errors. Here's an example crash dump:

```
Finalizing database at /tmp/codeql/dbs/<project>.
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f8d3bb5508b (sent by kill), pid=64262, tid=64303
#
# JRE version: OpenJDK Runtime Environment Temurin-17.0.8+7 (17.0.8+7) (build 17.0.8+7)
# Java VM: OpenJDK 64-Bit Server VM Temurin-17.0.8+7 (17.0.8+7, mixed mode, tiered, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# J 2971 c2 org.brotli.dec.Decode.decompress(Lorg/brotli/dec/State;)V (1729 bytes) @ 0x00007f8d3bb5508b [0x00007f8d3bb54f20+0x000000000000016b]
#
# Core dump will be written. Default location: Core dumps may be processed with "/usr/share/apport/apport -p%p -s%s -c%c -d%d -P%P -u%u -g%g -- %E" (or dumping to /usr/local/codeql-home/project/core.64262)
#
# An error report file with more information is saved as:
# /usr/local/codeql-home/project/hs_err_pid64262.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/adoptium/adoptium-support/issues
#
A fatal error occurred: Dataset import for /tmp/codeql/dbs/<project>/db-cpp failed with code 134.
```

The command we are running:

```
codeql database create /tmp/codeql/dbs/${CIRCLE_PROJECT_REPONAME} --overwrite --language=cpp --command='make sitl' --threads 1 --ram 35000
```

The strangest thing is that old commits that used to pass are now also failing. Any ideas as to what we could do to troubleshoot this?

**Comment by intrigus-lgtm** [comment]
(I don't work for GitHub, but this looks like an interesting issue)
Did you try running old commits with older versions of CodeQL?
Do you know which Java version was used for commits that used to pass?
Can you attach the `hs_err_pid64262.log` file?

**Comment by viktornordling** [comment]
> Did you try running old commits with older versions of CodeQL?

I've tried the most recent version (2.14.3) and (2.13.3), which is the version we've been on for a long while (which suddenly started failing). When I tried the old commits, it was using the same version (2.13.3) which it used to use when it worked.

> Do you know which Java version was used for commits that used to pass?
```
~/codeql/tools/linux64/java$ ./bin/java -version
openjdk version "17.0.7" 2023-04-18
OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7)
OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode)
```
Java seems to be bundled with codeql. I tried swapping
[hs_err_pid22225.log](https://github.com/github/codeql/files/12531745/hs_err_pid22225.log)
 that out with a new version of Oracle JDK, but it still crashed.

> Can you attach the hs_err_pid64262.log file?

Attached.

**Comment by viktornordling** [comment]
@intrigus-lgtm (or anyone else who sees this): any idea for things we could try?

**Comment by jbj** [comment]
The engineers who are best positioned to diagnose this problem have been on holiday, but they're returning this week. We're mystified that you're getting SIGSEGV inside pure Java code.

**Comment by viktornordling** [comment]
Thanks for your response @jbj! We are mystified too! 😸

**Comment by jbj** [comment]
We've updated the library for Brotli decoding in CodeQL CLI 2.14.4, which is due to come out this week. When you're able to test that, please let us know if the issue has gone away.

**Comment by intrigus-lgtm** [comment]
I'd suggest waiting for CodeQL CLI 2.14.4 and if it does not help, you can try these steps:
1. Does CodeQL run successfully for smaller projects?
2. Do you know how to run CodeQL with a "fastdebug" JVM (e.g. https://builds.shipilev.net/openjdk-jdk17/openjdk-jdk17-linux-x86_64-server-fastdebug-gcc10-glibc2.31.tar.xz)?
3. Can you run CodeQL with these additional Java options and attach the `app.log` file?
```
-Xlog:codecache*=trace,class+unload,class+load,os+thread,safepoint*,gc*,gc+ergo=trace,gc+age=trace,gc+phases=trace,gc+humongous=trace,jit+compilation=debug:file=app.log:level,tags,time,uptime,pid:filesize=104857600,filecount=5
```
This logs a lot of stuff, most of it probably being unnecessary. It is adapted from https://www.linkedin.com/pulse/javajvm-logs-monday-jvm-introduction-safepoints-krzysztof-%C5%9Blusarski. The important thing that this is tracing is everything related to **safepoints**.

You should be able to make CodeQL use these options by running CodeQL like this:
Instead of `codeql whatever command` run it like this:
```shell
codeql -J=-Xlog:codecache*=trace,class+unload,class+load,os+thread,safepoint*,gc*,gc+ergo=trace,gc+age=trace,gc+phases=trace,gc+humongous=trace,jit+compilation=debug:file=app.log:level,tags,time,uptime,pid:filesize=104857600,filecount=5 whatever command
```

My investigation:
We can take the hs_err file and disassemble the instruction around `pc` using a [small python script](https://gist.github.com/intrigus-lgtm/fa1b4b097579193c160ede487234af0e) I wrote.
We get this disassembly:
```
[...]
..>7f3e4d813f97:       4d 8b 9f 48 03 00 00    mov    r11, QWORD PTR [r15+0x348]
    7f3e4d813f9e:       41 83 c1 03             add    r9d, 0x3
    7f3e4d813fa2:       83 c6 03                add    esi, 0x3
    7f3e4d813fa5:       41 ff c1                inc    r9d
    7f3e4d813fa8:       ff c6                   inc    esi
-->7f3e4d813faa:       41 85 03                test   DWORD PTR [r11], eax
    7f3e4d813fad:       3b 1c 24                cmp    ebx, DWORD PTR [rsp]
    7f3e4d813fb0:       0f 8d 7a f5 ff ff       jge    0x7f3e4d813530
[...]
```
The crash happens at the position marked ("-->").
For someone that isn't deep into JVM land, this might look weird.
What is the point of executing a `test` instruction (a `test` instruction `and`-s the two operands together and discards the result, but updates the _flag_ registers) when the next instruction (`cmp`) immediately overwrites the _flag_ register?
But this is used as a way to implement [safepoints in Java](https://shipilev.net/jvm/anatomy-quarks/22-safepoint-polls/).
("..>" loads the address of the polling page from thread local storage (r15) and stores it in r11, while "-->" does the actual check)
If we look at register `r11` we can see that it holds the address `0x00007f3e58eb7000` which is the "Polling page: 0x00007f3e58eb7000" as per the hs_err file.
So this makes sense.

What doesn't make sense (to me) are two things:
1. The line "(sent by kill)" indicates that the JVM was killed _externally_ by doing `kill -SIGSEGV` which would be extremely weird to do (IMHO), especially with such a signal. This is maybe bogus and maybe the JVM misinterprets stuff here?
3. Why does the JVM crash? The address being accessed is the polling page, which is currently mapped as **unreadable** so that threads fault and run the intentionally installed signal handler of the JVM that will then bring the faulting thread to a **safe** **point**.
If the JVM wants to perform a safepoint (as the logs indicate "Event: 4.736 Executing VM operation: HandshakeAllThreads") it should be prepared to handle a SIGSEGV!
There are some events like "Event: 4.724 Protecting memory [0x00007f3e062fc000,0x00007f3e06300000] with protection modes 0" which look like they _arm_ the safe point page for a thread. If we crossreference the memory we can see that it is part of the "pool-1-thread-2" thread which is **blocked** as one would expect during a safe point.
If we do this for all other "Protecting memory" events, these correspond to "pool-1-thread[2-8]" and "C1 CompilerThread2" which are ALL blocked, **EXCEPT** "pool-1-thread-8" which is the crashing thread...
This is also weird, because the hs_err file explicitly says that we are not at a safepoint "VM state: not at safepoint (normal execution)"...

**Comment by intrigus-lgtm** [comment]
There is also another crash report that looks very similar to this one:
https://youtrack.jetbrains.com/issue/QD-3997

**Comment by viktornordling** [comment]
Thanks @jbj and @intrigus-lgtm!

@jbj, I think I will await that next release then.

@intrigus-lgtm, wow that's an impressive analysis. I keep getting back to this "killed externally" thing. Isn't it likely that the process is getting oom killed somehow? If it is, how would I find out? Unfortunately I don't have sudo on the circleci docker node that runs this (when I re-run with ssh), so the "normal" tricks I find online (like `sudo dmesg`) don't work. I think there should be a way to get sudo, but I haven't figured that out yet.

It's a good idea to try to run codeql on a smaller project, I'll try that.

**Comment by jbj** [comment]
[2.14.4](https://github.com/github/codeql-cli-binaries/releases/tag/v2.14.4) is released.

**Comment by viktornordling** [comment]
Same error with 2.14.4 unfortunately. This is probably something CircleCI / Docker related. I can run codeql on this codebase on my Linux laptop..

**Comment by smowton** [comment]
If you can reproduce in a local docker environment it would be interesting to inspect the `si_pid` member of the siginfo struct to find out who's sending SIGSEGV with `kill`. It sounds like something a virtualisation layer might do to reraise a signal it had intercepted but which it turned out it couldn't resolve (perhaps the JVM is reraising the signal on itself, having similarly hoped to intervene by expanding a stack or the like?)

I took a look briefly at whether the "sent by kill" could be a mistake -- I don't think so, as even though kernel-raised SIGSEGV signals overload the `si_code` field to tell you what sort of memory access fault they're describing, the values it fills in appear to be disjoint with the "what sort of signal is this" `SI_USER` etc values that the JVM is noticing and rendering as "sent by kill".

**Comment by intrigus-lgtm** [comment]
@smowton 
> If you can reproduce in a local docker environment it would be interesting to inspect the `si_pid` member of the siginfo struct to find out who's sending SIGSEGV with kill.

The hs_err file already contains this information and it does not make sense:
> siginfo: si_signo: 11 (SIGSEGV), si_code: 0 (SI_USER), si_pid: 0 (invalid), si_uid: 0


@viktornordling 
We clearly need a bigger hammer, so let's get one.
1. Download and replace the bundled JVM with a `fastdebug` variant:
    ```shell
    cd /usr/local/codeql-home/codeql/tools/linux64/ || exit 1 && \
    wget https://builds.shipilev.net/openjdk-jdk17/openjdk-jdk17-linux-x86_64-server-fastdebug-gcc10-glibc2.31.tar.xz && \
    tar xf openjdk-jdk17-linux-x86_64-server-fastdebug-gcc10-glibc2.31.tar.xz && \
    rm -rf java && \
    mv jdk java
   ```
2. Create a file `/tmp/gdb-commands` with this content:
```gdb
set logging file /tmp/gdb.log
set logging on
set logging redirect on
show logging
set breakpoint pending on
set confirm off
handle SIGSEGV nostop noprint pass
break VMError::report_and_die
commands
  p *(siginfo_t*)siginfo
  continue
end
run
```

Change your command
```
codeql database create /tmp/codeql/dbs/${CIRCLE_PROJECT_REPONAME} --overwrite --language=cpp --command='make sitl' --threads 1 --ram 35000
```
to this one:
```
gdb -batch -x /tmp/gdb-commands --args codeql database create /tmp/codeql/dbs/${CIRCLE_PROJECT_REPONAME} --overwrite --language=cpp --command='make sitl' --threads 1 --ram 35000
```
3. Attach the `/tmp/gdb.log` file to this issue.


5. Replace the JVM as in the previous first step.
6. Change your command
```
codeql database create /tmp/codeql/dbs/${CIRCLE_PROJECT_REPONAME} --overwrite --language=cpp --command='make sitl' --threads 1 --ram 35000
```
to this one:
```
strace -o /tmp/strace.log -k -i -e 'trace=!all' -f codeql database create /tmp/codeql/dbs/${CIRCLE_PROJECT_REPONAME} --overwrite --language=cpp --command='make sitl' --threads 1 --ram 35000
```
7. Attach the `/tmp/strace.log` file to this issue.

**Comment by smowton** [comment]
FWIW I tested and a process signalling itself (with kill(0, SIGSEGV)) gets si_pid = getpid() not zero. I also note the si_pid field is aliased via a union with the fields recording the address of the SEGV, which further supports the idea that this signal isn't really delivered via kill.

**Comment by viktornordling** [comment]
I think this is resolved. It turned out to be caused by the host machines for the CircleCI runners being under provisioned.

The weird thing is that no matter how little memory we gave codeql with the `--ram` parameter, it would still get killed.
We did notice that a new Java process is spawned during the "finalizing database" step, but as far as we could see that process would stay within the limits we had given it, so it's still not clear to us why it got killed.
